fv3gfs-util/.jenkins/test_util.sh

```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+set -e
+
+SCRIPT_DIR="$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+
+if [ "${target}" == "gpu" ] ; then
+    # we only run this on HPC
+    set +e
+    module load cray-python
+    module load pycuda
+    set -e
+fi
+
+# run tests
+echo "restoring cache"
+
+UTIL_DIR=$SCRIPT_DIR/..
+
+cache_key=v1-util-$($SCRIPT_DIR/checksum.sh $SCRIPT_DIR/test_util.sh $UTIL_DIR/requirements.txt $UTIL_DIR/requirements_gpu.txt $UTIL_DIR/constraints.txt)-$target
+
+$SCRIPT_DIR/cache.sh restore $cache_key
+
+echo "running tests"
+
+python3 -m venv venv
+. ./venv/bin/activate
+
+if [ "${target}" == "gpu" ] ; then
+    set +e
+    module unload cray-python
+    module unload pycuda
+    set -e
+    pip3 install -r $UTIL_DIR/requirements.txt -r $UTIL_DIR/requirements_gpu.txt -c $UTIL_DIR/constraints.txt -e $UTIL_DIR
+else
+    pip3 install -r $UTIL_DIR/requirements.txt -c $UTIL_DIR/constraints.txt -e $UTIL_DIR
```

Comment:
why do we still need the pip installs if we copy in the venv?

Reply:
Because the first time you run the script, the venv is empty. The way it is supposed to work is, this script should run identically if you delete the cache calls, but adding the cache lines makes it run faster while behaving the same.

Reply:
I see.
The only question I have is do we really gain much from copying the venv then?
Checking if all the requirements are fulfilled is about as fast as getting the wheels from somewhere, right? And either way we need a persistent storage somewhere. 
I'm not expecting an answer in absolute time, but I'm just curious if you have a gut feeling?

Reply:
> Checking if all the requirements are fulfilled is about as fast as getting the wheels from somewhere, right?

This is much faster in my experience, even when all the dependencies are wheels. But also it is very hard to ensure all the dependencies are wheels. Our current pace-fv3core-daint-setup plan last took 4 minutes 22 seconds to run, and the only thing it does other than install python packages appears to be linting.